k9mail/src/main/java/com/fsck/k9/activity/setup/outgoing/AccountSetupOutgoing.java

```diff
@@ -354,68 +347,13 @@ private void updateViewFromAuthType() {
-
-        if (isAuthTypeExternal && !hasConnectionSecurity) {
-
-            // Notify user of an invalid combination of AuthType.EXTERNAL & ConnectionSecurity.NONE
-            String toastText = getString(R.string.account_setup_outgoing_invalid_setting_combo_notice,
-                    getString(R.string.account_setup_incoming_auth_type_label),
-                    AuthType.EXTERNAL.toString(),
-                    getString(R.string.account_setup_incoming_security_label),
-                    ConnectionSecurity.NONE.toString());
-            Toast.makeText(this, toastText, Toast.LENGTH_LONG).show();
-
-            // Reset the views back to their previous settings without recursing through here again
-            OnItemSelectedListener onItemSelectedListener = mAuthTypeView.getOnItemSelectedListener();
-            mAuthTypeView.setOnItemSelectedListener(null);
-            mAuthTypeView.setSelection(mCurrentAuthTypeViewPosition, false);
-            mAuthTypeView.setOnItemSelectedListener(onItemSelectedListener);
-            updateViewFromAuthType();
-
-            onItemSelectedListener = mSecurityTypeView.getOnItemSelectedListener();
-            mSecurityTypeView.setOnItemSelectedListener(null);
-            mSecurityTypeView.setSelection(mCurrentSecurityTypeViewPosition, false);
-            mSecurityTypeView.setOnItemSelectedListener(onItemSelectedListener);
-            updateAuthPlainTextFromSecurityType(getSelectedSecurity());
-
-            mPortView.removeTextChangedListener(validationTextWatcher);
-            mPortView.setText(mCurrentPortViewSetting);
-            mPortView.addTextChangedListener(validationTextWatcher);
-
-            authType = getSelectedAuthType();
-            isAuthTypeExternal = (AuthType.EXTERNAL == authType);
-
-            connectionSecurity = getSelectedSecurity();
-            hasConnectionSecurity = (connectionSecurity != ConnectionSecurity.NONE);
-        } else {
-            mCurrentAuthTypeViewPosition = mAuthTypeView.getSelectedItemPosition();
-            mCurrentSecurityTypeViewPosition = mSecurityTypeView.getSelectedItemPosition();
-            mCurrentPortViewSetting = mPortView.getText().toString();
-        }
-
-        boolean hasValidCertificateAlias = mClientCertificateSpinner.getAlias() != null;
-        boolean hasValidUserName = Utility.requiredFieldValid(mUsernameView);
-
-        boolean hasValidPasswordSettings = hasValidUserName
-                && !isAuthTypeExternal
-                && Utility.requiredFieldValid(mPasswordView);
 
-        boolean hasValidExternalAuthSettings = hasValidUserName
-                && isAuthTypeExternal
-                && hasConnectionSecurity
-                && hasValidCertificateAlias;
+        presenter.revokeInvalidSettings(getSelectedAuthType(), getSelectedSecurity());
+        presenter.validateFields(mClientCertificateSpinner.getAlias(), mServerView.getText().toString(),
```

Comment:
the view should only tell the presenter "what is happening", but not "what to do".

the event in this case is that input changed, so this should fire something like `onInputChanged`. the presenter then decides what to do with that, i.e. verify the input, and updates the view accordingly.